Fix deprecations in nibabel and numpy

lukeje opened the pull request:
Both [nibabel](https://nipy.org/nibabel/reference/nibabel.spatialimages.html#nibabel.spatialimages.SpatialImage.get_affine) and [numpy](https://numpy.org/doc/stable/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated) have recently changed deprecation warnings to deprecation errors that break the installation process of gradunwarp. The first of these was already noted in issue #7. The changes here fix these deprecation errors.

I also made some changes to the description of the installation process and included a [requirements.txt](https://pip.pypa.io/en/stable/reference/requirements-file-format/) file to try and make the installation a little easier. In principle specific versions of packages should be specified in the requirements file, but I have not done this as I'm not sure what versions you test with.

These changes will potentially cause issues for backwards compatibility with python2.

Commit:
Replace deprecated nibabel.get_affine() with nibabel.affine

See here for more information:
https://nipy.org/nibabel/reference/nibabel.spatialimages.html#nibabel.spatialimages.SpatialImage.get_affine

Commit:
Replace deprecated numpy.float with numpy.float64

See here for more information:
https://numpy.org/doc/stable/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated

Commit:
Requirements file to enforce nibabel compatability

lukeje committed:
Add virtual environment installation option

mharms reviewed (commented):
This looks pretty straightforward (thanks @lukeje ).  What do you think @coalsont ?

mharms commented:
It would be good to know if this indeed is not compatible with Python 2.7, so that we can be clear on that.

coalsont commented on requirements.txt:
```diff
@@ -0,0 +1,3 @@
+nibabel>=4.0
```
It looks like the line adding the affine property dates back 9 years (from the tags, looks like version 2.0.0):

https://github.com/nipy/nibabel/commit/11151705c87d3c54bb3dc635df77ad8f25f72818#diff-bf1768d37a4ffcc36b8398d4d8b04b13ef1f492cfc371e84516eff208a9283b7R335

4.0.0 is the absolute newest release, from june 2022.  Should we loosen this version requirement so people are more likely to be able to use it without excess changes in an existing python environment?

lukeje replied:
It seems like the numpy deprecation was only fixed in nibabel version [3.2.1](https://github.com/nipy/nibabel/releases/tag/3.2.1). That version thus seems to be the earliest version which is compatible with the latest version of numpy. I will update the requirements file to match that.

Commit:
Minimal nibabel version compatible with latest numpy

Commit:
Make version name compliant with PEP440

pip has deprecated (https://github.com/pypa/pip/issues/8102) version names that do not comply with PEP440 (https://peps.python.org/pep-0440/).
Making this change also avoids what seems to be a bug in newer versions of setuptools which can mangle the shebang (#!) when installing gradient_unwarp.py using the fallback option used by pip for non-compliant packages.

Commit:
pip requirements file for python2

mharms commented on setup.py:
```diff
@@ -41,7 +41,7 @@ def configuration(parent_package='', top_path=None):
 
 setup(name='gradunwarp',
-      version = 'HCP-1.2.0',
+      version = '1.2.0+HCP',
```
@coalsont Should we make this version "1.2.1"?

coalsont replied:
The release versioning process should generally be separate from a PR.

mharms replied:
So, are you saying we'll just fix this later, as part of creating a new release?

coalsont replied:
Making a new release is a very different topic from fixing an incompatibility, and should not be conflated.  Also, we forked this repo, we didn't make it originally, so choosing to make a 1.2.1 version is not to be taken lightly.

coalsont replied:
On the other hand, the original repo never made a release, and we have already made several version numbers.  I think the "HCP" part of the version was in front intentionally, in order to detach it from the original repo versioning (which currently says `2.0_alpha`).  `HCP-1.2.0` does not mean "version 1.2.0, plus HCP's modifications" (which is what `1.2.0+HCP` looks like to me), it means "HCP's separate versioning scheme, 1.2.0".

mharms replied:
It was called `HCP-1.2.0` because that coincides with the release version on Github.  Not sure what standard caused @lukeje to change it, or even why a versions need to be included setup.py at all.  (Is that a python thing?)  To further complicate keeping everything in sync, there is a separate `version.txt` file as well we need to worry about.

coalsont replied:
So, if we want to be PEP compliant, we should either pretend the original repo didn't say `2.0_alpha` and suffer with any versioning weirdness if someone installs from the old repo, or to avoid that, either change the package name to include "HCP", use `3.0.0` (`2.1.0` doesn't leave much room if upstream is picked back up), or use an epoch like `1!1.2.0` (and we would have to keep using `1!<whatever>` forever).

coalsont replied:
We should probably just delete version.txt for simplicity.

coalsont replied:
I think we don't want to change the package name, and probably don't need to worry about people installing from the old repo.  So, next release can be plain `1.2.1`, and if we decide to revisit this later, we haven't taken away any options within PEP compliance.

coalsont replied:
Reading PEP440, it specifically says plus signs are not compliant.  So, it should probably just be `1.2.0` in this PR.

lukeje replied:
## Validity of `+` in PEP 440 version names
@coalsont You can see [here](https://peps.python.org/pep-0440/#local-version-identifiers) that PEP 440 allows adding a "local version identifier" after a `+` sign. You are right that it seems like this is intended to mean "version V plus the local modifications from site S", though. I would be fine with just dropping the identifier entirely if that's what you want to do.

## Deprecation of non-PEP 440 compliant names
You can see the pip deprecation notice here (using pip 22.3.1 with python 3.10.6):
```
$ pip install ./gradunwarp.source/ --force-reinstall
Processing ./gradunwarp.source
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: gradunwarp
  Building wheel for gradunwarp (setup.py) ... done
  Created wheel for gradunwarp: filename=gradunwarp-HCP_1.2.0-cp310-cp310-linux_x86_64.whl size=71375 sha256=7e494f2172e30b3dd8449cba5b14d3e410acf2b5d0e20545e8c8e15280eead24
  Stored in directory: /tmp/pip-ephem-wheel-cache-xc4qtw0c/wheels/20/43/d2/a1f7647295cb7995420aa5786ec680e5a7ef4c2c473841f7f3
  WARNING: Built wheel for gradunwarp is invalid: Metadata 1.2 mandates PEP 440 version, but 'HCP-1.2.0' is not
Failed to build gradunwarp
Installing collected packages: gradunwarp
  Attempting uninstall: gradunwarp
    Found existing installation: gradunwarp 1.2.0+hcp
    Uninstalling gradunwarp-1.2.0+hcp:
      Successfully uninstalled gradunwarp-1.2.0+hcp
  Running setup.py install for gradunwarp ... done
  DEPRECATION: gradunwarp was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. pip 23.1 will enforce this behaviour change. A possible replacement is to fix the wheel build issue reported above. Discussion can be found at https://github.com/pypa/pip/issues/8368
Successfully installed gradunwarp-HCP-1.2.0
```

For completeness, here is the list of installed packages in this virtual environment:
```
$ pip list
Package    Version
---------- ---------
gradunwarp HCP-1.2.0
nibabel    4.0.2
numpy      1.24.1
packaging  22.0
pip        22.3.1
scipy      1.9.3
setuptools 65.6.3
wheel      0.38.4
```
Unfortunately this legacy install method seems to run into another bug, wherein the shebang at the top of the gradient_unwarp.py becomes useless (as a relative path is not portable):
```bash
$ head gradunwarp.build/bin/gradient_unwarp.py -n1
#!python
```
I don't know where this bug comes from as it wasn't present when I tried installing gradunwarp on a different computer with python 3.9. I couldn't find anything specific on web forums about this issue, but there were hints that it could be because of the "legacy" installation method skipping pip's install tools and using setuptools instead.

If the version in setup.py is changed, then it can use pip's standard tools:
```
$ pip install ./gradunwarp.source/ --force-reinstall
Processing ./gradunwarp.source
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: gradunwarp
  Building wheel for gradunwarp (setup.py) ... done
  Created wheel for gradunwarp: filename=gradunwarp-1.2.0+hcp-cp310-cp310-linux_x86_64.whl size=71378 sha256=9cf8c3acb55c14062601ab0eb6009c4068505d0483c074be1984718c94d59a5f
  Stored in directory: /tmp/pip-ephem-wheel-cache-85j_0jl3/wheels/20/43/d2/a1f7647295cb7995420aa5786ec680e5a7ef4c2c473841f7f3
Successfully built gradunwarp
Installing collected packages: gradunwarp
  Attempting uninstall: gradunwarp
    Found existing installation: gradunwarp 1.2.0-HCP
    Uninstalling gradunwarp-1.2.0-HCP:
      Successfully uninstalled gradunwarp-1.2.0-HCP
Successfully installed gradunwarp-1.2.0+hcp
```
and gradient_unwarp.py then has the correct absolute path in the shebang:
```bash
$ head gradunwarp.build/bin/gradient_unwarp.py -n1
#!/home/luke/work/gradient-nonlinearity-correction-scripts/gradunwarp.build/bin/python
```

lukeje replied:
(Also, I can move the version renaming to a separate PR if you would like)

lukeje replied:
> We should probably just delete version.txt for simplicity.

To avoid redundancy you can read the version number into the [setup.py file](https://github.com/Washington-University/gradunwarp/blob/ce1d85c156e0ba202c30c788de90f34484e31e6d/setup.py) using something like
```python
with open("version.txt") as f:
    version = f.read()
    
setup(name='gradunwarp',
    version = version,
    ...
    )
```
But if you choose to do this then it should probably be a separate pull request (perhaps as part of the release version update).

coalsont replied:
I think we can leave it at `1.2.0+HCP`, while it doesn't really mean what "local build" is intended for, it may still be nice to keep the HCP string in the version info for clarity.

Commit:
Add python2 virtual environment installation instructions

Commit:
Include python2 and python3 requirements in the same file

coalsont commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-nibabel>=3.2.1
+nibabel >= 3.2.1 ; python_version > "3.0"
```
Is strictly greater the typical test for the python version?  Not that someone is likely to be running exactly python 3.0, but...

lukeje replied:
Oops... thanks for catching that. I'll fix it.

lukeje commented:
> It would be good to know if this indeed is not compatible with Python 2.7, so that we can be clear on that.

As @coalsont pointed out, the minimum version of nibabel is 2.0.0 for using the `affine` property rather than the `get_affine()` method. This seems to be sufficient to get the tool working for python2.7.

However nibabel 2.0.0 has the same deprecated numpy issue (using `numpy.float` rather than `numpy.float64`). This was fixed in nibabel 3.2.1, and so I would recommend having this as the minimum version for python3.

Commit:
Forgot that it should be less than or equal to python3.0...

glasserm commented:
Thanks Luke for your efforts in fixing these issues.  Do Tim and Mike feel that this pull request is ready for merge?

coalsont commented on README.md:
```diff
@@ -13,36 +13,76 @@ for use within the [HCP Minimal Preprocessing Pipelines][HCP Pipelines].
 
-#### Details
+### Install using a virtual environment
+If the [python virtual environment module](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments) is not already installed, then you may need to run
```
This might be confusing to users, as it may suggest that a virtualenv is required.  Can we add a sentence like "You may optionally choose to install gradunwarp into a python virtual environment in order to avoid changing your existing python setup"?

coalsont commented on README.md:
````diff
@@ -13,36 +13,76 @@ for use within the [HCP Minimal Preprocessing Pipelines][HCP Pipelines].
-```
-sudo apt-get install python3-numpy python3-pip
+```bash
+sudo apt install python3-numpy python3-pip
````
Is it a good idea to have python3-numpy in here, when numpy is in the requirements.txt?

lukeje replied:
I just checked, and it doesn't seem to really matter. 
If `numpy` and `scipy` are already installed (e.g. because `apt install python3-numpy` was run), then the `-r requirements.txt` step just says e.g.
```
Requirement already satisfied: numpy in /usr/lib/python3/dist-packages
```
and if they are not then it installs the latest versions.

coalsont commented:
I ran a quick test, and the code seems to work.

coalsont committed:
Note that virtual environment is optional

coalsont commented:
Merging this now because it works, feel free to comment on the remaining documentation question when you have the time.

mharms commented:
What, if anything, is remaining in your mind?
